File: src/org/cappuccino_project/ide/intellij/plugin/contributor/ObjJFindUsagesProvider.java
package org.cappuccino_project.ide.intellij.plugin.contributor;

import com.intellij.lang.HelpID;
import com.intellij.lang.cacheBuilder.DefaultWordsScanner;
import com.intellij.lang.cacheBuilder.WordsScanner;
import com.intellij.lang.findUsages.FindUsagesProvider;
import com.intellij.psi.PsiElement;
import com.intellij.psi.tree.TokenSet;
import org.cappuccino_project.ide.intellij.plugin.lexer.ObjJLexer;
import org.cappuccino_project.ide.intellij.plugin.psi.*;
import org.cappuccino_project.ide.intellij.plugin.psi.types.ObjJTypes;
import org.cappuccino_project.ide.intellij.plugin.psi.utils.ObjJHasContainingClassPsiUtil;
import org.cappuccino_project.ide.intellij.plugin.psi.utils.ObjJPsiImplUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class ObjJFindUsagesProvider implements FindUsagesProvider {


    @Nullable
    @Override
    public WordsScanner getWordsScanner() {
        return new DefaultWordsScanner(
                new ObjJLexer(),
                TokenSet.create(ObjJTypes.ObjJ_SELECTOR, ObjJTypes.ObjJ_VARIABLE_NAME, ObjJTypes.ObjJ_CLASS_NAME, ObjJTypes.ObjJ_FUNCTION_NAME),
                TokenSet.create(ObjJTypes.ObjJ_SINGLE_LINE_COMMENT, ObjJTypes.ObjJ_BLOCK_COMMENT),
                TokenSet.create(ObjJTypes.ObjJ_INTEGER, ObjJTypes.ObjJ_STRING_LITERAL, ObjJTypes.ObjJ_DECIMAL_LITERAL, ObjJTypes.ObjJ_BOOLEAN_LITERAL)
        );
    }

    @Override
    public boolean canFindUsagesFor(
            @NotNull
                    PsiElement psiElement) {
        return  psiElement instanceof ObjJSelector ||
                psiElement instanceof ObjJVariableName ||
                psiElement instanceof ObjJClassName ||
                psiElement instanceof ObjJFunctionName;
    }

    @Nullable
    @Override
    public String getHelpId(
            @NotNull
                    PsiElement psiElement) {
        return HelpID.FIND_OTHER_USAGES;
    }

    @NotNull
    @Override
    public String getType(
            @NotNull
                    PsiElement psiElement) {
        if (psiElement instanceof ObjJSelector) {
            return "method Selector";
        } else if (psiElement instanceof ObjJVariableName) {
            return "variable";
        } else if (psiElement instanceof ObjJClassName) {
            return "class";
        } else if (psiElement instanceof ObjJFunctionName) {
            return "function";
        }
        return "";
    }

    @NotNull
    @Override
    public String getDescriptiveName(
            @NotNull
                    PsiElement psiElement) {
        final String containingClassOrFileName = ObjJHasContainingClassPsiUtil.getContainingClassOrFileName(psiElement);
        return ObjJPsiImplUtil.getDescriptiveText(psiElement) + " in " + containingClassOrFileName;
    }

    @NotNull
    @Override
    public String getNodeText(
            @NotNull
                    PsiElement psiElement, boolean b) {
        return ObjJPsiImplUtil.getDescriptiveText(psiElement);
    }
}

File: src/cappuccino/ide/intellij/plugin/contributor/ObjJFindUsagesProvider.kt
package cappuccino.ide.intellij.plugin.contributor

import cappuccino.ide.intellij.plugin.lexer.ObjJLexer
import cappuccino.ide.intellij.plugin.psi.ObjJClassName
import cappuccino.ide.intellij.plugin.psi.ObjJFunctionName
import cappuccino.ide.intellij.plugin.psi.ObjJSelector
import cappuccino.ide.intellij.plugin.psi.ObjJVariableName
import cappuccino.ide.intellij.plugin.psi.types.ObjJTypes
import cappuccino.ide.intellij.plugin.psi.utils.ObjJHasContainingClassPsiUtil
import cappuccino.ide.intellij.plugin.psi.utils.ObjJPsiImplUtil
import com.intellij.lang.HelpID
import com.intellij.lang.cacheBuilder.DefaultWordsScanner
import com.intellij.lang.cacheBuilder.WordsScanner
import com.intellij.lang.findUsages.FindUsagesProvider
import com.intellij.psi.PsiElement
import com.intellij.psi.tree.TokenSet

class ObjJFindUsagesProvider : FindUsagesProvider {

    override fun getWordsScanner(): DefaultWordsScanner {
        return DefaultWordsScanner(
                ObjJLexer(),
                TokenSet.create(ObjJTypes.ObjJ_SELECTOR, ObjJTypes.ObjJ_VARIABLE_NAME, ObjJTypes.ObjJ_CLASS_NAME, ObjJTypes.ObjJ_FUNCTION_NAME),
                TokenSet.create(ObjJTypes.ObjJ_SINGLE_LINE_COMMENT, ObjJTypes.ObjJ_BLOCK_COMMENT),
                TokenSet.create(ObjJTypes.ObjJ_INTEGER, ObjJTypes.ObjJ_STRING_LITERAL, ObjJTypes.ObjJ_DECIMAL_LITERAL, ObjJTypes.ObjJ_BOOLEAN_LITERAL)
        )
    }

    override fun canFindUsagesFor(
            psiElement: PsiElement): Boolean {
        return psiElement is ObjJSelector ||
                psiElement is ObjJVariableName ||
                psiElement is ObjJClassName ||
                psiElement is ObjJFunctionName
    }

    override fun getHelpId(
            psiElement: PsiElement): String? {
        return HelpID.FIND_OTHER_USAGES
    }

    override fun getType(
            psiElement: PsiElement): String {
        return when (psiElement) {
            is ObjJSelector -> "method Selector"
            is ObjJVariableName -> "variable"
            is ObjJClassName -> "class"
            is ObjJFunctionName -> "function"
            else -> ""
        }
    }

    override fun getDescriptiveName(
            psiElement: PsiElement): String {
        val containingClassOrFileName = ObjJHasContainingClassPsiUtil.getContainingClassOrFileName(psiElement)
        return ObjJPsiImplUtil.getDescriptiveText(psiElement) + " in " + containingClassOrFileName
    }

    override fun getNodeText(
            psiElement: PsiElement, b: Boolean): String {
        return ObjJPsiImplUtil.getDescriptiveText(psiElement) ?: psiElement.text
    }
}
